### PR TITLE
fix: reinstate form template validation in API

### LINKED ISSUE #2184: form API doesn't validate form templates

### PR DESCRIPTION
**Pull request by opqdonut**
fixes #2184

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue

## Documentation
- [x] update changelog if necessary

## Testing
- [x] complex logic is unit tested
- [x] valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [ ] new tasks are created for pending or remaining tasks
- [x] no critical TODOs left to implement

### CODE REVIEW

File: test/clj/rems/api/test_forms.clj [hunk 1]
```diff
@@ -397,29 +411,26 @@
                                 :field/privacy :public}]}]
     (testing "creating"
       (testing "invalid request"
-        (letfn [(fail-request [command]
-                  (let [response (-> (request :post "/api/forms/create")
-                                     (authenticate api-key user-id)
-                                     (json-body command)
-                                     handler)]
-                    (or (= 400 (:status response))
-                        (not (get-in response [:body :success])))))]
```
Comment (opqdonut) [inline test/clj/rems/api/test_forms.clj:406]: The bug was here: `:body` needs to be read & parsed. It's a stream and thus never contains `:success`.

File: test/clj/rems/api/test_forms.clj [hunk 2]
```diff
@@ -397,29 +411,26 @@
...
-          (is (fail-request (assoc-in command [:form/fields 1 :field/privacy] nil)) "invalid value")
-          (is (fail-request (assoc-in command [:form/fields 1 :field/privacy] :does-not-exist)) "invalid value")
-          (is (fail-request (assoc-in command [:form/fields 0 :field/privacy] :private)) "privacy not supported")))
```
Comment (opqdonut) [inline test/clj/rems/api/test_forms.clj:409]: This was never a feature of form validation, I added a test case to rems.common.form that demonstrates this

File: src/cljc/rems/common/form.cljc [hunk 3]
```diff
@@ -245,6 +248,15 @@
                      (assoc-in [:form/fields 0 :field/placeholder :fi] ""))]
         (is (empty? (validate-form-template form languages)))))
 
+    (testing "unused values are not validated"
+      (let [form (assoc form :form/fields
+                        [{:field/title {:en "en" :fi "fi"}
+                          :field/type :header
+                          :field/max-length "-3"
+                          :field/privacy :invalid
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:256]: See here how `:field/privacy` can be anything if it's not supported for the field type

File: test/clj/rems/api/test_end_to_end.clj [hunk 4]
```diff
@@ -367,7 +373,8 @@
                                                   :form/title "e2e"
                                                   :form/fields [{:field/type :text
                                                                  :field/title {:en "text field"
-                                                                               :fi "tekstikenttä"}
+                                                                               :fi "tekstikenttä"
+                                                                               :sv "textfält"}
```
Comment (foxlynx) [inline test/clj/rems/api/test_end_to_end.clj:377]: A bit inconsistent that above all the titles are "text field" for all languages and here translated. If there aren't two many in the file maybe change all to translated ones?

File: src/cljc/rems/common/form.cljc [hunk 5]
```diff
@@ -157,6 +157,9 @@
       (empty? (:visibility/values visibility))
       {:field/visibility {:visibility/values :t.form.validation/required}}
 
+      (not (apply distinct? (:visibility/values visibility)))
+      {:field/visibility {:visibility/values :t.form.validation/invalid-value}}
```
[the remark this answers is not in the document]
Comment (opqdonut) [reply src/cljc/rems/common/form.cljc:161]: it's detecting cases with duplicated values like `:visibility/values ["yes" "yes"]`

this validation isn't strictly necessary, but we had a test case for it, so I added it

File: test/clj/rems/api/test_end_to_end.clj (hunk 4 above)
Comment (opqdonut) [reply test/clj/rems/api/test_end_to_end.clj:377]: fair enough :)

File: src/cljc/rems/common/form.cljc (hunk 5 above)
[the remark this answers is not in the document]
Comment (foxlynx) [reply src/cljc/rems/common/form.cljc:161]: How about checking against a list of approved visibility values rather? This doesn't prohibit other bogus values than duplicates.

[the remark this answers is not in the document]
Comment (opqdonut) [reply src/cljc/rems/common/form.cljc:161]: that check is below, line 163

[the remark this answers is not in the document]
Comment (foxlynx) [reply src/cljc/rems/common/form.cljc:161]: Ahh, true.

File: src/cljc/rems/common/form.cljc (hunk 3 above)
Comment (opqdonut) [reply src/cljc/rems/common/form.cljc:256]: ... I ended up just implementing the missing feature

File: src/cljc/rems/common/form.cljc [hunk 6]
```diff
@@ -168,20 +171,30 @@
       nil {:field/visibility {:visibility/type :t.form.validation/required}}
       {:field/visibility {:visibility/type :t.form.validation/invalid-value}})))
 
+(defn- validate-not-present [field key]
+  (when (contains? field key)
+    {key :unsupported}))
```
Comment (opqdonut) [inline src/cljc/rems/common/form.cljc:176]: this error isn't localized because it shouldn't ever be generated by the UI

**Pull request merged**